### PR TITLE
Handle updates of Service and Check definitions

### PR DESCRIPTION
**Pull request by diptanu**
This PR handles syncing of Service and Check definitions more precisely

### CODE REVIEW

File: client/consul.go [hunk 1]
```diff
@@ -75,119 +90,174 @@ func (c *ConsulClient) Register(task *structs.Task, allocID string) error {
...
+
+func (c *ConsulService) performSync(agent *consul.Agent) {
+	var consulServices map[string]*consul.AgentService
+	var consulChecks map[string]*consul.AgentCheck
+	var err error
+
+	// Remove the tracked services which tasks no longer references
+	for serviceId, ts := range c.trackedServices {
+		if !ts.IsServiceValid() {
+			c.logger.Printf("[INFO] consul: Removing service: %s since the task doesn't have it anymore", ts.service.Name)
```
Comment (dadgar) [inline client/consul.go]: I would make this DEBUG

File: client/consul.go [hunk 2]
```diff
@@ -75,119 +90,174 @@ func (c *ConsulClient) Register(task *structs.Task, allocID string) error {
...
+		}
+	}
+}
+
+func (c *ConsulService) performSync(agent *consul.Agent) {
+	var consulServices map[string]*consul.AgentService
+	var consulChecks map[string]*consul.AgentCheck
+	var err error
+
+	// Remove the tracked services which tasks no longer references
```
Comment (dadgar) [inline client/consul.go]: references -> reference

File: client/consul_test.go [hunk 3]
```diff
@@ -93,8 +93,74 @@ func TestInvalidPortLabelForService(t *testing.T) {
...
+					DynamicPorts: []structs.Port{{"db", 20413}},
+				},
+			},
+		},
+	}
+	c.Register(&task, "1")
+	task.Services = []*structs.Service{}
+
+	c.performSync(c.client.Agent())
+	if len(c.trackedServices) != 0 {
```
Comment (dadgar) [inline client/consul_test.go]: You should check that this is == 1 before

**Comment by dadgar** [comment]
LGTM but I think we can simplify the logic. It is a bit hard to follow right now. I think we should have a task have a list of services which has a list of checks. Then you can do a two way diff on each of those objects recursively.

**Comment by diptanu** [comment]
The data structures are already like that, and I agree the logic could be simplified.

**Comment by diptanu** [comment]
We will have a followup PR for simplifying the code which tracks checks and services which are registered.
